src/PooledArrays.jl

```diff
@@ -304,7 +304,62 @@ Base.findall(pdv::PooledVector{Bool}) = findall(convert(Vector{Bool}, pdv))
 ##
 ##############################################################################
 
-function Base.map(f, x::PooledArray{T,R}) where {T,R<:Integer}
+"""
+    map(f, x::PooledArray; pure::Bool=false)
+
+Transform `PooledArray` `x` by applying `f` to each element.
+
+If `pure=true` then `f` is applied to each element of pool of `x`
+exactly once (even if some elements in pool are not present it `x`).
+This will typically be much faster when the proportion of unique values
+in `x` is small.
+"""
+function Base.map(f, x::PooledArray; pure::Bool=false)
+    pure && return _map_pure(f, x)
+    length(x) == 0 && return PooledArray(collect(Base.Generator(f, x)))
+    v1 = f(x[1])
+    T = typeof(v1)
+    invpool = Dict(v1 => one(eltype(x.refs)))
+    pool = [v1]
+    labels = similar(x.refs)
+    labels[1] = 1
+    nlabels = 1
+    return _map_notpure(f, x, 2, invpool, pool, labels, nlabels)
+end
+
+function _map_notpure(f, xs::PooledArray, start,
+                      invpool::Dict{T,I}, pool::Vector{T},
+                      labels::Array{I}, nlabels::Int) where {T, I<:Integer}
+    for i in start:length(xs)
+        vi = f(xs[i])
+        Ti = typeof(vi)
+        lbl = get(invpool, vi, zero(I))
+        if lbl !== zero(I)
+            labels[i] = lbl
+        else
+            if nlabels == typemax(I) || !(Ti === T)
+                I2 = nlabels == typemax(I) ? _widen(I) : I
```

Comment:
This is another difficult decision: this makes `map` type-unstable, which could have dramatic consequences for performance inside package code that was not written with `PooledArray`s in mind. On the other hand, if we don't widen we have to throw an error, which isn't great either.

An alternative could be to widen directly to `UInt32`, so that `map` can only return two array types: either with the same reftype as the input, or with `UInt32` as a reftype. That way small union optimizations would be used.

Yet another alternative would be to return a plain `Vector` if the reftype is too small to hold all values. This could make sense given that if you create new levels, you're most likely returning random or unique values.

Reply:
This is indeed a potential problem. However, note that `PooledArray` constructor is already type unstable also. I propose to leave things as is. I add an additional comment to the docstrings of `map` and `PooledArray` to cover this.

Reply:
Yeah, but when you call the constructor you know you're creating a `PooledArray`, so you can deal with the type instablity if needed. On the contrary, if the call to `map` happens in package code, this PR may introduce performance regressions that users cannot do anything about.

Given that currently we don't even support returning a larger number of levels than in the input, we could just throw an error for now, and deal with reftype widening later depending on whether it turns out to be a problem in practice or not. It's always easier to turn an error into something else than the reverse.

Reply:
> Given that currently we don't even support returning a larger number of levels than in the input

We do not, because for pure functions it is not possible to have a larger number of levels. So this error would be never thrown.

> this PR may introduce performance regressions that users cannot do anything about.

Yes - I was thinking about this. However, the alternative is to throw an error.

The third option - as you commented is to use input type or maximum widening immediately (`Int64` or `UInt64`). Then we would have an union of two concrete types. Do you think we should go this way?

Reply:
> We do not, because for pure functions it is not possible to have a larger number of levels. So this error would be never thrown.

Yes that's what I mean.

> Yes - I was thinking about this. However, the alternative is to throw an error.
> 
> The third option - as you commented is to use input type or maximum widening immediately (`Int64` or `UInt64`). Then we would have an union of two concrete types. Do you think we should go this way?

Yeah, let's do that. We could use `UInt32` as an intermediate step, but returning three different types would be a bit complex.